### PR TITLE
Implement custom connection pool

### PR DESCRIPTION
**Pull request by prudhvi**
Fixes https://github.com/squareup/spirit/issues/180

### CODE REVIEW

File: pkg/dbconn/connpool.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+type ConnPool struct {
+	sync.Mutex
+	config *DBConfig
+	conns  []*sql.Conn
+}
+
+// NewRRConnPool creates a pool of transactions which have already
+// had their read-view created in REPEATABLE READ isolation.
+func NewRRConnPool(ctx context.Context, db *sql.DB, count int, config *DBConfig) (*ConnPool, error) {
+	checksumTxns := make([]*sql.Conn, 0, count)
```
Comment (morgo) [inline pkg/dbconn/connpool.go]: We can rename this, since it is checksum agnostic.

File: pkg/dbconn/connpool.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		return
+	}
+	p.Lock()
+	defer p.Unlock()
+	p.conns = append(p.conns, trx)
+}
+
+// Close closes all transactions in the pool.
+func (p *ConnPool) Close() error {
+	fmt.Println("ABout to close all connections ####")
```
Comment (morgo) [inline pkg/dbconn/connpool.go]: I don't mind if you want to Debug log this. To do so though, you would need to plumb a logger through in dbConfig.

File: pkg/dbconn/dbconn.go [hunk 3]
```diff
@@ -35,6 +33,37 @@ func NewDBConfig() *DBConfig {
...
+	_, err = conn.ExecContext(ctx, "SET lock_wait_timeout=?", config.LockWaitTimeout)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 func standardizeTrx(ctx context.Context, trx *sql.Tx, config *DBConfig) error {
 	_, err := trx.ExecContext(ctx, "SET time_zone='+00:00'")
 	if err != nil {
```
Comment (morgo) [inline pkg/dbconn/dbconn.go]: I think (actually I hope) this func can also be removed? We now only have to standardize connections

File: pkg/dbconn/dbconn_test.go [hunk 4]
```diff
@@ -4,12 +4,9 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"github.com/stretchr/testify/assert"
```
Comment (morgo) [inline pkg/dbconn/dbconn_test.go]: I think this is wrong, and probably is triggering a linter. Non built-in imports go last.

File: pkg/checksum/checker.go [hunk 5]
```diff
@@ -28,7 +28,8 @@ type Checker struct {
 	concurrency int
 	feed        *repl.Client
 	db          *sql.DB
-	trxPool     *dbconn.TrxPool
+	regularPool *dbconn.ConnPool // non snapshot connection
+	pool        *dbconn.ConnPool // RR snapshot connections
```
Comment (morgo) [inline pkg/checksum/checker.go]: I think we should maybe flip these around and keep the regular pool as `pool` so it's the same in all components? You can also remove `db` and use `pool.DB()`.

File: pkg/row/copier.go [hunk 6]
```diff
@@ -97,7 +96,7 @@ func NewCopier(db *sql.DB, tbl, newTable *table.TableInfo, config *CopierConfig)
 }
 
 // NewCopierFromCheckpoint creates a new copier object, from a checkpoint (copyRowsAt, copyRows)
-func NewCopierFromCheckpoint(db *sql.DB, tbl, newTable *table.TableInfo, config *CopierConfig, lowWatermark string, rowsCopied uint64, rowsCopiedLogical uint64) (*Copier, error) {
+func NewCopierFromCheckpoint(db *dbconn.ConnPool, tbl, newTable *table.TableInfo, config *CopierConfig, lowWatermark string, rowsCopied uint64, rowsCopiedLogical uint64) (*Copier, error) {
 	c, err := NewCopier(db, tbl, newTable, config)
```
Comment (morgo) [inline pkg/row/copier.go]:
```suggestion
func NewCopierFromCheckpoint(pool *dbconn.ConnPool, tbl, newTable *table.TableInfo, config *CopierConfig, lowWatermark string, rowsCopied uint64, rowsCopiedLogical uint64) (*Copier, error) {
	c, err := NewCopier(pool, tbl, newTable, config)
```

File: pkg/migration/runner.go [hunk 7]
```diff
@@ -534,19 +539,20 @@ func (r *Runner) postCutoverCheck(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	checker, err := checksum.NewChecker(r.db, oldTable, cutoverTable, r.replClient, &checksum.CheckerConfig{
+	ssPoolForChecksum, err := dbconn.NewPoolWithConsistentSnapshot(ctx, r.db, 1, r.pool.DBConfig(), r.logger)
+	if err != nil {
+		return err
+	}
```
Comment (prudhvi) [inline pkg/migration/runner.go:546]: ssPool currently only gets initialized in checker.Run method so have to add option in CheckerConfig and supply from runner here for this step cc @morgo

**Pull request closed without merging**